### ISSUE TITLE
Add version (v3) to package name in Java

### ISSUE THREAD

**Issue by marcinwyszynski** [google/protobuf#1717]
TL;DR I would like to propose that package `com.google.protobuf` becomes `com.google.protobuf.v3` in Java. It is conceivable that one wants/needs to use both library versions simultaneously.

`proto3` is not backwards compatible with `proto2`, at least in Java. I noticed that when I tried to use `proto3` to serialise data to be put into HBase. The thing is that HBase at version `1.2.1` (latest) internally uses version `2.5.0`of the library and one of messages it defines overrides `writeTo` method - which is marked `final` as of `3.0.0-beta`.

Regardless of the reasoning behind modifying autogenerated code this essentially means that I can't use `proto3` with HBase which for me is a major bummer since I want to be able to use gRPC as well as serve the Ruby client which does not support `proto2` 😱.

One way or another, instead of trying to provide backwards compatibility which is a hard target to hit I would like to suggest that `proto3` is explicitly namespaced as to allow both versions to coexist in one app.

**Comment by xfxyjwf** [comment]
This is a known issue (reported quite frequently by hadoop/hbase users). We discussed and decided to work on fixing up v3.0.0 to make it compatible with 2.5.0 generated code for ease of migration. We are against renaming the package name because that will effectively create two protobuf Java APIs and impose a high migration cost for existing protobuf users to move to the new version.

We are still working on it and will probably soon publish a Java specific version that can be used as a bridge version between 2.5.0 and 3.0.0.

**Comment by marcinwyszynski** [comment]
![go4it](https://cloud.githubusercontent.com/assets/571786/16413610/347da55a-3d33-11e6-89f3-7ddf6ef044d4.jpg)

**Comment by amardeep** [comment]
What is the status of this?

**Comment by xfxyjwf** [comment]
@amardeep I have created a 2.7.0 branch for Java. If you got binary compatibility problems when upgrading from 2.5.0 to 3.0.0 directly, 2.7.0 can be used as a migration step.

**Comment by silvanheller** [comment]
@xfxyjwf I have a problem specifically with `com.google.protobuf.LiteralByteString` which HBase uses up to the newest version (1.x) and will only disappear in 2.0 but disappears in Proto 3. I'm not finding a protobuf-java:2.7 published anywhere and 2.6.1 isn't compatible with V3 generated Code. This means there is still no way to use proto 3 and HBase together, correct?

**Comment by dingwa** [comment]
@silvanheller, from what I understand. Protobuf 3.x is not compatible with HBase (protobuf 2.5.0). I specifically asked the below question with HBase in mind. We have the same problem

https://groups.google.com/forum/m/#!searchin/protobuf/Scott$20sue/protobuf/AwJlDT8ZWjQ

**Comment by silvanheller** [comment]
@dingwa your link is broken for me

**Comment by xfxyjwf** [comment]
@silvanheller Unfortunately LiteralByteString is a protobuf private symbol and there is no way for us to preserve compatibility for those who are accessing protobuf private symbols. The only solution is for HBASE to remove their reference to protobuf private symbols, and then you should be able to use it with protobuf 3.0.0-beta-4 which is compatible with 2.5/2.6 if only protobuf public APIs are used.

**Comment by dingwa** [comment]
@silvanheller apologies. Link updated

**Comment by markthegrea** [comment]
I am assuming that the 2.7 jar is what we need to bridge this gap.  Is there any place to download it?  I can not seem to get the project to open in Eclipse/Maven.
